### ISSUE TITLE
nn.TransformerLayer

### ISSUE THREAD

**Issue by aszlam** [pytorch/pytorch#21876]
A generic pytorch transformer layer 

## 🚀 Feature
input: 

a 2-dim Tensor [x_{00}, x_{01},  ... x_{0m_1}, x_{10}, ... x_{1m_1}, ... x_{km_k}]; each x is a vector.

and either (or both, but they should match):
* a 1-dim Tensor of offsets [o_0, ..., o_k ]
* or a list of square 2-dim Tensors [ A_{m_0}, .... A_{m_k}] of attention masks

output: 
a 2-dim Tensor [h_{00}, h_{01},  ... h_{0m_1}, h_{10}, ... h_{1m_1}, ... h_{km_k}]

## Motivation

This is a generic transformer formulation, and can be used for text, vision, graphs, etc.

**Comment by aszlam** [comment]
@cpuhrsch @zhangguanheng66 @soumith @imisra @siddk

**Comment by zhangguanheng66** [comment]
@aszlam Thanks for the contribution. A great feature to extend the applicability of the transformer model in other domains.

**Comment by zwx8981** [comment]
Is there any plan to provide a tutorial of showing an example of using Transformer layer as an alternative of RNN for seq2seq task such as machine translation? @cpuhrsch @zhangguanheng66 @soumith @imisra @siddk

**Comment by zhangguanheng66** [comment]
@zwx8981 there is and will land soon. https://github.com/pytorch/examples/pull/555

**Comment by zhangguanheng66** [comment]
The TransformerBag plus MultiheadAttentionBag will add more flexibility for the input sequences. For the sequences with different lengths, no padding is required, and mask/offsets are used for the attention mechanism. Though it’s a pleasant to have such kind of capability, it requires some fundamental changes on the current modules (both nn.MultiheadAttention and nn.Transformer ). We have to re-design the modules and re-write them from scratch. We identify it as a nice feature to have and are open to some contributors for the implementation. The existing modules have been already applied for some non-text based sequences (like PySpeech). Users just need to pad the sequences before sending to the current attention mechanism.
